ddtrace/internal/telemetry/metrics.py

```diff
@@ -0,0 +1,131 @@
+        """
+        name: metric name
+        common: set to True if a metric is common to all tracers, false if it is python specific
+        interval: field set for gauge and rate metrics, any field set is ignored for count metrics (in secs)
+        """
+        self.name = name
+        self.common = common
+        self.interval = interval
+        self._roll_up_interval = interval
+        self.namespace = namespace
+        self._points = []  # type: List[Tuple[int, float]]
+        self._tags = tags  # type: MetricTagType
+        self._count = 0.0
+
+    @property
+    def id(self):
+        """
+        https://www.datadoghq.com/blog/the-power-of-tagged-metrics/#whats-a-metric-tag
+        """
+        return self.name + str(self._tags)
+
+    def __hash__(self):
+        return self.id
+
+    @abc.abstractmethod
+    def add_point(self, value=1.0):
+        # type: (float) -> None
+        """adds timestamped data point associated with a metric"""
+        pass
+
+    def set_tags(self, tags):
+        # type: (Dict) -> None
+        """sets a metrics tag"""
+        self._tags = tags
+
+    def set_tag(self, name, value):
+        # type: (str, str) -> None
+        """sets a metrics tag"""
+        self._tags[name] = value
+
+    def to_dict(self):
+        # type: () -> Dict
+        """returns a dictionary containing the metrics fields expected by the telemetry intake service"""
+        return {
+            "host": self.HOST_NAME,
+            "name": self.name,
+            "type": self.metric_type,
+            "common": self.common,
+            "interval": int(self.interval) if self.interval else None,
+            "points": self._points,
+            "tags": self._tags,
+        }
+
+
+class CountMetric(Metric):
+    """
+    A count type adds up all the submitted values in a time interval. This would be suitable for a
+    metric tracking the number of website hits, for instance.
+    """
+
+    metric_type = "count"
+
+    def add_point(self, value=1.0):
+        # type: (float) -> None
+        """adds timestamped data point associated with a metric"""
+        timestamp = int(time.time())
+        self._points.append((timestamp, float(value)))
+
+
+class GaugeMetric(Metric):
+    """
+    A gauge type takes the last value reported during the interval. This type would make sense for tracking RAM or
+    CPU usage, where taking the last value provides a representative picture of the host’s behavior during the time
+    interval. In this case, using a different type such as count would probably lead to inaccurate and extreme values.
+    Choosing the correct metric type ensures accurate data.
+    """
+
+    metric_type = "gauge"
+
+    def add_point(self, value=1.0):
+        # type: (float) -> None
+        """adds timestamped data point associated with a metric"""
+        timestamp = int(time.time())
+        self._points = [(timestamp, float(value))]
```

Comment:
Looks like this replaces the entire `_points` array with a single-element array. Why is that?

Reply:
cause GaugeMetric we want to store the last measure we take